### ISSUE TITLE
Separation of feature files

### ISSUE THREAD

**Issue by santhoshtr** [smc/fonts#3]
I started a branch to experiment with the separate feature files for our fonts. See https://github.com/smc/fonts/tree/featurefile

Pros:
1. Feature file resides in a plain text file - helping us to easily manage patches about features alone without generating not-so-easy to comprehend fontforge patches.
2. Separating glyph shape changes and feature changes
3. Potential reuse
4. Annotated feature file - adding comments about bug fixes and explanation for non-trivial feature definitions make font maintenance easier. 

Cons:
Font development workflow change. Suppose I wanted to fix a bug with rendering. To debug, I need the sfd with feature files applied. If I do changes in fontforge, it is sfd that changes and not the feature file. I need to manually export the changed feature file to .fea file. Or one need to make the changes in fea directly. Or the midway is fix in sfd and port it to .fea file.

If we want to fix some positioning fixes, we need graphical tool. Then also we need to make the .fea and sfd in sync.

In short, we need a clever workflow management here to make the font developers life easy. To make the changes we do in sfd using fontforge or handediting the .fea seamlessly and integrating that to make builds without loosing any fixes, suggestions welcome.

I noticed an issue with roundtrip - sfd->extract feature file -> apply feature file-> sfd. For Meera font, a markclass named chandrakkala was not exported and merging was failing while applying .fea later. I fixed it by handediting .fea temporarily, but anybody noticed this kind of issues?

ccying: @kavyamanohar @rajeeshknambiar @aravindavk @pravins @smc

**Comment by santhoshtr** [comment]
Corrected the URL as https://github.com/smc/fonts/tree/featurefile

**Comment by pravins** [comment]
On 14 December 2013 12:13, Santhosh Thottingal notifications@github.comwrote:

> In short, we need a clever workflow management here to make the font
> developers life easy. To make the changes we do in sfd using fontforge or
> handediting the .fea seamlessly and integrating that to make builds without
> loosing any fixes, suggestions welcome.

If you are aware about Font development on Windows with MS VOLT. If is very
much same process. Font designer designs shape in Fontographer and later
font developer/programmer use MS VOLT to import .ttf file and write
features/import .vtp file into it.

I am happy we are slowly moving on that line with some pros and cons as
santhosh pointed out.

> I noticed an issue with roundtrip - sfd->extract feature file -> apply
> feature file-> sfd. For Meera font, a markclass named chandrakkala was not
> exported and merging was failing while applying .fea later. I fixed it by
> handediting .fea temporarily, but anybody noticed this kind of issues?
> 
> ccying: @kavyamanohar https://github.com/kavyamanohar @rajeeshknambiarhttps://github.com/rajeeshknambiar
> @aravindavk https://github.com/aravindavk @pravinshttps://github.com/pravins
> @smc https://github.com/smc

Nope, i did not found this issue. It will be good if you can report issues
in fontforge upstream [1] for better debugging . Do keep me in cc.

Regards,
Pravin Satpute
1. https://github.com/fontforge/fontforge/issues/new

**Comment by santhoshtr** [comment]
I tried to find a way where files are not accidentally overwritten in the workflow. As per commit 6f2f414a70be9e80f636ec78c1fbbcd9f4f8ccdd I tried to see if there is a diff between exported fea file and fea file existing, and ask which version to use. The logic works fine, but came across an issue

If I export a feature file, say FEA1 and apply that to sfd, export again(call it FEA2). FEA1 and FEA2 are supposed to be same. But that is not the case, I see quite a lot of diff between FEA1 and FEA2. The diff is not functional change, but lines are swapped/interchanged. Because of this above logic fails.

Will report this in fontforge upstream.

**Comment by aravindavk** [comment]
How about maintaining featurefile only for GSUB rules. (Considering gsub
rules not change too often and easy to maintain the rules in featurefile).
We can expect only featurefile patch for gsub rules.

But positional rules changes looks very challenging, comments in the
featurefile will not be maintained when we export the featurefile.

On Sat, Dec 21, 2013 at 2:00 PM, Santhosh Thottingal <
notifications@github.com> wrote:

> I tried to find a way where files are not accidentally overwritten in the
> workflow. As per commit 6f2f414https://github.com/smc/fonts/commit/6f2f414a70be9e80f636ec78c1fbbcd9f4f8ccddI tried to see if there is a diff between exported fea file and fea file
> existing, and ask which version to use. The logic works fine, but came
> across an issue
> 
> If I export a feature file, say FEA1 and apply that to sfd, export
> again(call it FEA2). FEA1 and FEA2 are supposed to be same. But that is not
> the case, I see quite a lot of diff between FEA1 and FEA2. The diff is not
> functional change, but lines are swapped/interchanged. Because of this
> above logic fails.
> 
> Will report this in fontforge upstream.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/smc/fonts/issues/3#issuecomment-31059281
> .

## 

Regards
Aravinda | ಅರವಿಂದ
http://aravindavk.in

**Comment by santhoshtr** [comment]
No longer valid since we dont edit opentype rules in fontforge and SFD does not contain opentype tables
